src/jni_core/src/lib.rs

```diff
@@ -607,27 +610,127 @@ extern "system" fn Java_com_risingwave_java_binding_Binding_iteratorGetTimestamp
+                    },
+                    jvalue {
+                        i: value.0.minute() as i32,
+                    },
+                    jvalue {
+                        i: value.0.second() as i32,
+                    },
+                    jvalue {
+                        i: value.0.nanosecond() as i32,
+                    },
+                ],
+            )?
+            else {
+                return Err(BindingError::from(jni::errors::Error::MethodNotFound {
+                    name: "of".to_string(),
+                    sig: sig.into(),
+                }));
+            };
+            Ok(timestamp_obj)
+        }
+    })
+}
+
+#[no_mangle]
+extern "system" fn Java_com_risingwave_java_binding_Binding_iteratorGetTimestamptzValue<'a>(
+    env: EnvParam<'a>,
+    pointer: Pointer<'a, JavaBindingIterator<'a>>,
+    idx: jint,
+) -> JObject<'a> {
+    execute_and_catch(env, move |env: &mut EnvParam<'_>| {
+        let value = pointer
+            .as_ref()
+            .datum_at(idx as usize)
+            .unwrap()
+            .into_timestamptz()
+            .to_datetime_utc();
+
+        let utc = {
+            let cls = env.find_class(gen_class_name!(java.time.ZoneOffset))?;
+            env.get_static_field(&cls, "UTC", gen_jni_type_sig!(java.time.ZoneOffset))?
+                .l()?
+        };
+
+        let sig = gen_jni_sig!(java.time.OffsetDateTime of(int year, int month, int dayOfMonth, int hour, int minute, int second, int nanoOfSecond, java.time.ZoneOffset offset));
+
+        let (timestamptz_class_ref, constructor) = pointer
+            .as_ref()
+            .class_cache
+            .timestamptz_ctor
+            .get_or_try_init(|| {
+                let cls = env.find_class(gen_class_name!(java.time.OffsetDateTime))?;
+                let init_method = env.get_static_method_id(&cls, "of", sig)?;
+                Ok::<_, jni::errors::Error>((env.new_global_ref(cls)?, init_method))
+            })?;
+        unsafe {
+            let JValueOwned::Object(timestamptz_obj) = env.call_static_method_unchecked(
+                <&JClass<'_>>::from(timestamptz_class_ref.as_obj()),
+                *constructor,
+                ReturnType::Object,
+                &[
+                    jvalue { i: value.year() },
+                    jvalue {
+                        i: value.month() as i32,
+                    },
+                    jvalue {
+                        i: value.day() as i32,
+                    },
+                    jvalue {
+                        i: value.hour() as i32,
+                    },
+                    jvalue {
+                        i: value.minute() as i32,
+                    },
+                    jvalue {
+                        i: value.second() as i32,
+                    },
+                    jvalue {
+                        i: value.nanosecond() as i32,
+                    },
+                    jvalue { l: *utc },
```

Comment:
It seems that the [rw timestamptz](https://github.com/risingwavelabs/risingwave/blob/main/src/common/src/types/timestamptz.rs#L36) type only stores the microseconds?

Reply:
Oh I get the idea now. I think maybe it's better to use [static OffsetDateTime	ofInstant(Instant instant, ZoneId zone)](https://docs.oracle.com/javase/8/docs/api/java/time/OffsetDateTime.html#ofInstant-java.time.Instant-java.time.ZoneId-). We can create the `Instant` from the raw 64 bit timestamp instead of converting to a datetime and then fetch the year month and data.

And also the `UTC` can be stored as a global ref in the cache to avoid keep fetching it.